src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs

```diff
@@ -0,0 +1,549 @@
+{
+    public override void M<T>() { }
+
+    public void Test()
+    {
+        this.M<int>();
+        this.M<E>();
+    }
+}
+public enum E
+{
+}", references: new[] { reference }).VerifyDiagnostics(
+                // (8,9): error CS0315: The type 'int' cannot be used as type parameter 'T' in the generic type or method 'B.M<T>()'. There is no boxing conversion from 'int' to 'System.Enum'.
+                //         this.M<int>();
+                Diagnostic(ErrorCode.ERR_GenericConstraintNotSatisfiedValType, "this.M<int>").WithArguments("B.M<T>()", "System.Enum", "T", "int").WithLocation(8, 9));
+        }
+
+        [Fact]
+        public void EnumConstraint_EnforcedInInheritanceChain_Upwards_Source()
+        {
+            CreateStandardCompilation(@"
+public abstract class A
+{
+    public abstract void M<T>();
+}
+public class B : A
+{
+    public override void M<T>() where T : System.Enum { }
+}").VerifyDiagnostics(
+                // (8,33): error CS0460: Constraints for override and explicit interface implementation methods are inherited from the base method, so they cannot be specified directly
+                //     public override void M<T>() where T : System.Enum { }
+                Diagnostic(ErrorCode.ERR_OverrideWithConstraints, "where").WithLocation(8, 33));
+        }
+
+        [Fact]
+        public void EnumConstraint_EnforcedInInheritanceChain_Upwards_Reference()
+        {
+            var reference = CreateStandardCompilation(@"
+public abstract class A
+{
+    public abstract void M<T>();
+}").EmitToImageReference();
+
+            CreateStandardCompilation(@"
+public class B : A
+{
+    public override void M<T>() where T : System.Enum { }
+}", references: new[] { reference }).VerifyDiagnostics(
+                // (4,33): error CS0460: Constraints for override and explicit interface implementation methods are inherited from the base method, so they cannot be specified directly
+                //     public override void M<T>() where T : System.Enum { }
+                Diagnostic(ErrorCode.ERR_OverrideWithConstraints, "where").WithLocation(4, 33));
+        }
+
+        [Fact]
+        public void EnumConstraint_ResolveParentConstraints()
+        {
+            var comp = CreateStandardCompilation(@"
+public enum MyEnum
+{
+}
+public abstract class A<T>
+{
+    public abstract void F<U>() where U : System.Enum, T;
+}
+public class B : A<MyEnum>
+{
+    public override void F<U>() { }
+}");
+
+            Action<ModuleSymbol> validator = module =>
+            {
+                var method = module.GlobalNamespace.GetTypeMember("B").GetMethod("F");
+                var constraintTypeNames = method.TypeParameters.Single().ConstraintTypes().Select(type => type.ToTestDisplayString());
+
+                AssertEx.SetEqual(new[] { "System.Enum", "MyEnum" }, constraintTypeNames);
+            };
+
+            CompileAndVerify(comp, sourceSymbolValidator: validator, symbolValidator: validator);
+        }
+    }
+}
```

Comment:
Some more test ideas, if they are not already covered:
- cross-language scenario: calling an enum-constrained method from VB
- cross-assembly scenario: calling an enum-constrained method defined in a first compilation from a second compilation (image reference and compilation reference). Verify that symbols loaded from PE have the constraint as expected.

Reply:
* Cross language scenario: already covered: `EnumConstraint_FromCSharp`
* Cross assembly scenario: using symbols from a different assembly defined in tests starting with `EnumConstraint_Reference*`, and validating the symbols in tests starting with `EnumConstraint_IsReflectedinSymbols*`

I'll add tests for the rest.

---
In reply to: [162502034](https://github.com/dotnet/roslyn/pull/24199#discussion_r162502034) [](ancestors = 162502034)

Reply:
@jcouv correct me if I'm wrong, but we are not supporting operators right now, as the emitted IL will differ based on the underlying type of the IL. So operators cannot be applied to generic types still.

---
In reply to: [162510809](https://github.com/dotnet/roslyn/pull/24199#discussion_r162510809) [](ancestors = 162510809,162502034)